bayesian/utility/db_gateway.py

```diff
@@ -98,6 +107,100 @@ def get_batch_ca_data(cls, ecosystem: str, packages) -> dict:
+        elapsed_time = time.time() - started_at
+        logger.info("It took %s to fetch vuln results from Gremlin.", elapsed_time)
+
+        return response.json()
+
+    @classmethod
+    def get_package_details(cls, ecosystem: str, packages) -> dict:
+        """Get vulnerabilities for given packages."""
+        logger.debug('Executing get_package_details')
+
+        payload = {
+            'gremlin': cls.get_package_query,
+            'bindings': {
+                'ecosystem': ecosystem,
+                'packages': packages
+            }
+        }
+        started_at = time.time()
+        response = post(url=gremlin_url, data=json.dumps(payload))
+        response.raise_for_status()
+        elapsed_time = time.time() - started_at
+        logger.info("It took %s to fetch pckg results from Gremlin.", elapsed_time)
+
+        return response.json()
+
+    @classmethod
+    def filter_vulnerable_packages(cls, vulnerabilities: list, package_version_map: dict) -> dict:
+        """Filter vulnerable package based on timestamp in pseudo version."""
+        logger.debug('Executing filter_vulnerable_packages')
+
+        filter_vulnerabilities = {}
+        gh = GithubUtils()
+        for vuln in vulnerabilities:
+            package_name = vuln['package_name'][0]
+            if gh._is_commit_date_in_vuln_range(
+               gh.extract_timestamp(
+                   package_version_map[package_name]), vuln['vuln_commit_date_rules'][0]):
+                if package_name not in filter_vulnerabilities:
+                    filter_vulnerabilities[package_name] = []
+                filter_vulnerabilities[package_name].append(vuln)
+
+        return filter_vulnerabilities
+
+    @classmethod
+    def get_batch_ca_data_for_pseudo_version(cls, ecosystem: str, packages) -> dict:
+        """Component analyses batch call only for pseudo version applicable for golang."""
+        logger.debug('Executing get_batch_ca_data_for_pseudo_version')
+        started_at = time.time()
+
+        # Build unique package list and package -> version map.
+        filter_packages = set()
+        package_version_map = {}
+        for pckg in packages:
+            package_name = pckg['name'].split('@')[0]
+            filter_packages.add(package_name)
+            package_version_map[package_name] = pckg['version']
+
+        vuln_response = GraphAnalyses.get_vulnerabilities_for_packages(
+            ecosystem, list(filter_packages))
+        vulnerabilities = GraphAnalyses.filter_vulnerable_packages(
+            vuln_response.get('result', {}).get('data', []), package_version_map)
+        pckg_response = GraphAnalyses.get_package_details(
+            ecosystem, list(vulnerabilities.keys()))
+
+        # Merge the package and vunlerability data into response.
+        data = []
+        for pckg in pckg_response.get('result', {}).get('data', []):
+            data.append({
+                'package': pckg,
+                'version': {},
```

Comment:
Should the version field populated using package_version_map?, else why it is been kept empty?

Reply:
Pseudo based packages don't have version node in graph db. Also we don't need this in CA, So to explicitly indicate that Pseudo packages don't have 'version', I have set it to empty dict.

Reply:
The CA response should have a version right?, even for pseudo version?

Reply:
In response version is filled by 'given_version' that is read from request. We don't read version from Graph as of now.